base/sort.jl

```diff
@@ -652,13 +669,203 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, a::PartialQuickSort,
     return v
 end
 
+# This is a stable least significant bit first radix sort.
+#
+# That is, it first sorts the entire list by the last CHUNK_SIZE bits, then by the second to
+# last CHUNK_SIZE bits, and so on. Stability means that it will not reorder two elements
+# that compare equal. This is essential so that the order introduced by earlier,
+# less significant passes is preserved by later passes.
+#
+# Each pass divides the input into 2^CHUNK_SIZE == MASK+1 buckets. To do this, it
+#  * counts the number of entries that fall into each bucket
+#  * uses those counts to compute the indices to move elements of those buckets into
+#  * moves elements into the computed indices in the swap array
+#  * switches the swap and working array
+#
+# In the case of an odd number of passes, the returned list will === the input list t,
+# not v. This is one of the many reasons radix_sort! is not exported.
+function radix_sort!(v::AbstractVector{U}, lo::Integer, hi::Integer, bits::Unsigned,
+               ::Val{CHUNK_SIZE}, t::AbstractVector{U}) where {U <: Unsigned, CHUNK_SIZE}
+    # bits is unsigned and CHUNK_SIZE is a compile time constant for performance reasons.
+    MASK = UInt(1) << CHUNK_SIZE - 0x1
+    counts = Vector{UInt}(undef, MASK+2)
+
+    @inbounds for shift in 0:CHUNK_SIZE:bits-1
+
+        # counts[2:MASK+2] will store the number of elements that fall into each bucket.
+        # if CHUNK_SIZE = 8, counts[2] is bucket 0x00 and counts[257] is bucket 0xff.
+        counts .= 0
+        for k in lo:hi
+            x = v[k]                  # lookup the element
+            i = (x >> shift)&MASK + 2 # compute its bucket's index for this pass
+            counts[i] += 1            # increment that bucket's count
+        end
+
+        counts[1] = lo                # set target index for the first bucket
+        cumsum!(counts, counts)       # set target indices for subsequent buckets
+        # counts[1:MASK+1] now stores indices where the first member of each bucket
+        # belongs, not the number of elements in each bucket. We will put the first element
+        # of bucket 0x00 in t[counts[1]], the next element of bucket 0x00 in t[counts[1]+1],
+        # and the last element of bucket 0x00 in t[counts[2]-1].
+
+        for k in lo:hi # Is this iteration slower than it could be?
```

Comment:
What does this question mean? Either make it more explicit or drop it?

Reply:
When profiling I found that the actual iteration step seems to be taking longer than it should (perhaps because it is the only branch). I tried hacking in a [duff's device ](https://en.wikipedia.org/wiki/Duff%27s_device) but (as expected) that didn't change anything. At this point, I'm going to answer the question with "seems pretty decent to me" and take it out of the source code.

Reply:
Done